bin/main

```diff
@@ -0,0 +1,182 @@
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+    entry
+  end
+
+  #   [4,5,3,2].my_each_with_index {|val,index| puts "index: #{index} for #{val}" if val < 10}
+  #   output:
+  #   index: 0 for 4
+  #   index: 1 for 5
+  #   index: 2 for 3
+  #   index: 3 for 2
+
+  def my_select
+    return to_enum unless block_given?
+
+    i = 0
+    arr = []
+    while i < size
+      arr << self[i] if yield(self[i])
+      i += 1
+    end
+    arr
+  end
+  #   puts [1,2,3,4,5,6].my_select(&:even?)
+  #   output:
+  #   2
+  #   4
+  #   6
+
+  def my_all?(pattern = nil)
+    unless block_given?
+      return my_all? { |obj| obj } unless pattern
+
+      if pattern.class == Regexp
+        my_all? { |x| pattern.match(x) }
+      elsif pattern.class == Class
+        my_all? { |x| x.is_a? pattern }
+      else
+        my_all? { |x| x === pattern }
+      end
+    end
+    my_each { |x| return false unless yield x }
+    true
+  end
+    # puts %w[ant bear cat].my_all? { |word| word.length >= 8 }
+  #   output:
+  #   false
+
+  def my_any?(pattern = nil)
+    unless block_given?
+      return my_any? { |obj| obj } unless pattern
+
+      if pattern.class == Regexp
+        my_any? { |x| pattern.match(x) }
+      elsif pattern.class == Class
+        my_any? { |x| x.is_a? pattern }
+      else
+        my_any? { |x| x === pattern }
+      end
+    end
+    my_each { |x| return true if yield x }
+    false
+  end    
+    # puts %w[ant bear cat].my_any? { |word| word.length >= 8 }
+    #   output:
+    #   false
+
+  def my_none?(pattern = nil)
+    if block_given?
+      !my_any? { |x| yield x }
+    elsif pattern
+      !my_any?(pattern)
+    else
+      my_each { |x| return false if x }
+      true
+    end
+  end
+    # puts  ["ant", "bear", "cat"].my_none? { |word| word.length == 5 }
+  #   output:
+  #   true
+
+  def my_count(xander = nil)
+    count = 0
+    if block_given?
+      my_each { |e| count += 1 if yield e }
+    elsif xander
+      my_each { |e| count += 1 if e == xander }
+    else
+      count = length
+    end
+    count
+  end
+  #   ary = [1, 2, 4, 2]
+  #   puts ary.my_count               #=> 4
+  #   puts ary.my_count(2)            #=> 2
+  #   puts ary.my_count{ |x| x%2==0 } #=> 3
+
+  def my_map
+    return to_enum :my_map unless block_given?
+
+    i = 0
+    arr = []
+
+    while i < size
+      array2 << if block_given?
+                  yield(self[i])
+                else
+                  proc.call(self[i])
+                end
+      i += 1
+    end
+    arr
+  end
+
+  def my_inject(initial = nil, sym = nil)
+    if block_given?
+      my_each { |item| initial = yield initial, item }
+    else
+      unless sym
+        raise 'No symbol nor block given' unless initial.class == Symbol
+
+        sym = initial
+        initial = self[0]
+      end
+      my_each do |item|
+        next if initial == item
+
+        initial = initial.send sym, item
+      end
+    end
+    initial
+  end
```

Comment:
What happens when I pass a `proc` as an argument? Did you check? I got 

> NoMethodError:
>         undefined method `+' for nil:NilClass

Also, the result when passed an array is wrong. You may refer to [this api doc](https://apidock.com/ruby/Enumerable) if you need help understanding how these methods should work.